### PR TITLE
Add guidance on private module conventions

### PR DESCRIPTION
**Pull request by codeinthehole**
Which is an extension to the existing convenience import convention.

### CODE REVIEW

File: python.md [hunk 1]
```diff
@@ -1031,6 +1031,21 @@ A useful pattern is to import the "public" objects from a package into its
 `__init__.py` module to make life easier for calling code. This does need to be
 done with care though - here's a few guidelines:
 
+#### Prefix module names with an underscore
+
+If the `__init__.py` defines the _public_ API of your package, it makes sense to
+indicate that the other package modules are _private_ and shouldn't be imported
+directly by callers.
```
Comment (seddonym) [inline python.md:1038]:
```suggestion
directly by callers external to the parent package.
```

File: python.md [hunk 2]
```diff
@@ -1031,6 +1031,21 @@ A useful pattern is to import the "public" objects from a package into its
...
+
+Do this by using an underscore prefix, eg:
+
+```txt
+mypackage/
+    __init__.py  # Public API
+    _foo.py
+    _bar.py
+```
```
Comment (JackThorp) [inline python.md:1047]: So is it correct to say that if `__init__.py` defines the public API then *all* other modules in such a package should be private?

Comment (codeinthehole) [reply python.md:1047]: I think so. I can't think of a reason why not.

Comment (seddonym) [reply python.md:1047]: Personally I think not. You might have a subpackage which exposes a public interface.

Comment (codeinthehole) [reply python.md:1047]: > Personally I think not. You might have a subpackage which exposes a public interface.

But would you mix that with convenience imports in `__init__.py`? It seems more straightforward to pick one approach or the other.

Comment (seddonym) [reply python.md:1047]: That's a good point, I agree and hadn't thought about that before. If you're using the private module notation, there should be one canonical entry point for the outside world.

I reckon the key point here is about canonical entry points, and the private module naming is just a means of achieving that. I've done an edit of your version which might get this across - feel free to use or ignore: https://github.com/octoenergy/conventions/pull/59

Comment (codeinthehole) [reply python.md:1047]: Good stuff - I'll merge your version.

**Comment by codeinthehole** [comment]
Superseded by #59

**Pull request closed without merging**